qemu/tests/qemu_guest_agent.py

```diff
@@ -818,6 +818,247 @@ def gagent_check_reboot_shutdown(self, test, params, env):
+        except guest_agent.VAgentCmdError as detail:
+            if not re.search('guest-file-open has been disabled',
+                             str(detail)):
+                self.test.fail("This is not the desired information: "
+                               "('%s')" % str(detail))
+        else:
+            self.test.fail("guest-file-open command shouldn't succeed "
+                           "for freeze FS.")
+        finally:
+            self.gagent.fsthaw()
+
+        error_context.context("After thaw fs, try to operate guest"
+                              " file.", logging.info)
+        ret_handle = int(self.gagent.guest_file_open(tmp_file, mode="a+"))
+        content = "hello world\n"
+        self.gagent.guest_file_write(ret_handle, content)
+        self.gagent.guest_file_flush(ret_handle)
+        self.gagent.guest_file_seek(ret_handle, 0, 0)
+        self._read_check(ret_handle, "hello world")
+
+    @error_context.context_aware
+    def gagent_check_guest_file(self, test, params, env):
+        """
+        Test guest agent commands "guest-file-open/write/flush/
+        seek/read/close".
+
+        Test steps:
+        1) set file related cmd to qemu-ga white list(linux guest only)
+        2) basic test:
+          a. create file in guest and write "hello world" to it.
+          b. open and read file with mode "r".
+          c. close file.
+        3) do child function test
+        8) clear the temp file in guest
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environmen.
+        """
+        session = self._get_session(params, self.vm)
+        self._open_session_list.append(session)
+        error_context.context("Change guest-file related cmd to white list.")
+        self._change_bl(session)
+
+        try:
+            error_context.context("Basic guest-file cmd test.",
+                                  logging.info)
+            # read guest file which are already created in guest.
+            logging.info("Create file in guest.")
+            tmp_file = self._create_tmp_file()
+            content = "hello world"
+            session.cmd("echo %s > %s" % (content, tmp_file))
+
+            logging.info("Open and read the guest file.")
+            ret_handle = int(self.gagent.guest_file_open(tmp_file))
+            self._read_check(ret_handle, content)
+
+            logging.info("Close the opened file.")
+            self.gagent.guest_file_close(ret_handle)
+
+            # run the child function
+            test_type = params["test_type"]
+            chk_type = "_guest_file_%s" % test_type
+            logging.info("chk_type is %s." % chk_type)
+            if hasattr(self, chk_type):
+                file_func = getattr(self, chk_type)
+                logging.info("function is %s." % file_func)
+                file_func(session)
+                if ret_handle:
```

Comment:
Well, you don't get the ret_handle from the function, you actually get it from L#1037, that's not what we expected.

Reply:
Yes, I think again, it's better not to handle self.gagent.guest_file_close(ret_handle) after file_function.It's better to make file_fun flexible,such as new function which don't return ret_handle

Reply:
@xiagao The problem is with current implementation you do not perform close action for every sub test.

Reply:
I know that, I will modify it if you agree with my solution

Reply:
@xiagao Sure, you might close the files inside the functions.